### PR TITLE
Enabling DeepSpeed to support parameter-efficient finetuning of very large models on single GPU (11B+ parameters)

### PR DESCRIPTION
**Pull request by Raldir**
*Description of changes:*

- Enabling Deepspeed in AutoGluon, by adapting the associated [pytorch-lightning strategy](https://pytorch-lightning.readthedocs.io/en/stable/advanced/model_parallel.html#deepspeed-zero-stage-3). Setting the strategy to `deepspeed_stage_3_offload` now allows to fine-tune models such as T0pp (11B parameters) with parameter-efficient finetuning methods, such as IA^3, on a single A10G GPU (23GB memory), by offloading optimizer states to the CPU. Particularly suited to few-shot learning, where training is generally very quick and the effect of offloading is relatively minor (see PR #2032). Some hacky workarounds are currently in place and indicated in code, since pytorch-lightning has some limitations in its use.
- Now intermediate checkpoints only store weights that are updated during training, i.e. `requires_grad=True`. Added a flag `standalone` to `fit()` to indicate whether the final checkpoint should consist of the entire model's weights or only trained weights. In the case of T0pp and fine-tuning with (IA)^3, instead of needing 40GB per checkpoint only 4.2MB are needed. 
- Fixed gradient_checkpointing for T-Few, works correctly now.
- Adding seed  argument to `_predict()` to ensure reproducibility.
- Fixed some style/design issues regarding T-Few.

TODOs:
- Add zero-shot mode for T0pp.
- Add Deepspeed inference mode with custom kernels and quantization for improved inference speed and memory consumption, see [here](https://www.deepspeed.ai/tutorials/inference-tutorial/).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/constants.py [hunk 1]
```diff
@@ -137,6 +137,12 @@
 IA3_BIAS = "ia3_bias"
 IA3_NORM = "ia3_norm"
 
+# DeepSpeed constants (for now gather/reduce size are constants for single GPU)
+DEEPSPEED_OFFLOADING = "deepspeed_stage_3_offload"
+DEEPSPEED_MIN_PL_VERSION = "1.7.7"
+DEEPSPEED_ALLGATHER_SIZE = 1e9
+DEEPSPEED_ALLREDUCE_SIZE = 1e9
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/constants.py:144]: We can put the `ALLGATHER_SIZE` and `ALLREDUCE_SIZE` as configurable options in https://github.com/awslabs/autogluon/blob/master/multimodal/src/autogluon/multimodal/configs/environment/default.yaml .

File: multimodal/src/autogluon/multimodal/optimization/deepspeed.py [hunk 2]
```diff
@@ -0,0 +1,976 @@
+# Slightly adapted file of pytorch_lightning.strategies.deepspeed to adjust for additional memory requirements when running deepspeed inference.
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/optimization/deepspeed.py:1]: We can also summarize about what we have changed.

File: multimodal/src/autogluon/multimodal/optimization/lit_module.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def __init__(
         custom_metric_func: Callable = None,
         test_metric: Optional[torchmetrics.Metric] = None,
         efficient_finetune: Optional[str] = None,
-        trainable_param_names: Optional[List[str]] = None,
+        trainable_param_names: Optional[List] = [],
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/optimization/lit_module.py:44]: We should still use `None` as the default value. Also, should it be `List[str]`?

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/optimization/lit_module.py:44]: We can cast None to `[]` later .

File: multimodal/src/autogluon/multimodal/utils/checkpoint.py [hunk 4]
```diff
@@ -53,6 +61,70 @@ def average_checkpoints(
     return avg_state_dict
 
 
+class AutoMMModelCheckpointIO(pl.plugins.CheckpointIO):
+    """
+    Class that customizes how checkpoints are saved. Saves either the entire the entire model or only parameters that have been explicitly updated during training. The latter reduces memory footprint substentially when training very large models with parameter-efficient finetuning methods.
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/utils/checkpoint.py:66]: Nit: "the entire the entire" --> "the entire".

File: multimodal/src/autogluon/multimodal/utils/checkpoint.py [hunk 5]
```diff
@@ -53,6 +61,70 @@ def average_checkpoints(
...
+    Class is based on pl.plugins.TorchCheckpointIO.
+    """
+
+    def __init__(self, trainable_param_names, model_name_to_id):
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/utils/checkpoint.py:70]: Need to add docstring of `trainable_param_names` and `model_name_to_id`.

File: multimodal/src/autogluon/multimodal/optimization/utils.py [hunk 6]
```diff
@@ -535,7 +535,7 @@ def apply_layerwise_lr_decay(
     lr_decay: float,
     weight_decay: float,
     efficient_finetune: Optional[str] = None,
-    trainable_param_names: Optional[List[str]] = None,
+    trainable_param_names: Optional[List] = [],
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/optimization/utils.py:538]: Similar to the previous line. We should use `None` as default.

**Comment by sxjscience** [comment]
I noticed that there seems to an installation timeout issue associated mmcv-full. 

Error message:

```
 [2022-10-12T00:17:40.619Z] + mim install mmcv-full --default-timeout 120
[2022-10-12T00:17:42.038Z] Looking in links: https://download.openmmlab.com/mmcv/dist/cu102/torch1.12.0/index.html
[2022-10-12T00:17:42.878Z] Collecting mmcv-full
[2022-10-12T00:17:43.107Z]   Downloading https://download.openmmlab.com/mmcv/dist/cu102/torch1.12.0/mmcv_full-1.6.2-cp38-cp38-manylinux1_x86_64.whl (33.4 MB)
[2022-10-12T00:19:00.281Z]      ━━━━━━━━━━                               8.8/33.4 MB 262.1 kB/s eta 0:01:34
```

Will it be related to the timeout? @bryanyzhu 
https://github.com/awslabs/autogluon/blob/f8c0ede595093c13011fe6c50b1c7ffa8b9c1575/.github/workflow_scripts/env_setup.sh#L61

File: multimodal/src/autogluon/multimodal/data/process_text.py [hunk 7]
```diff
@@ -357,8 +357,8 @@ def build_one_token_sequence_from_text(
             if col_name == CHOICES_IDS:
                 answer_ids = self.tokenizer(
                     col_text,
-                    return_tensors="pt",
```
[the remark this answers is not in the document]
Comment (Raldir) [reply multimodal/src/autogluon/multimodal/data/process_text.py:360]: Thanks! :) Other point on the list was the `trainable_param_names` argument, which was an artefact then but now has proper functionality.

**Review by sxjscience — APPROVED** [review]
LGTM!! Thanks! Let’s add a tutorial afterwards.

File: .github/workflow_scripts/env_setup.sh [hunk 8]
```diff
@@ -58,7 +58,7 @@ function install_multimodal {
     # launch different process for each test to make sure memory is released
     python3 -m pip install --upgrade pytest-xdist
     python3 -m pip install --upgrade -e multimodal/[tests]
-    mim install mmcv-full --default-timeout 120
```
Comment (zhiqiangdon) [inline .github/workflow_scripts/env_setup.sh:61]: @FANGAreNotGnu added this flag in this [PR](https://github.com/awslabs/autogluon/pull/2196) as suggested by @bryanyzhu . I guess it is to handle the timeout issue when installing mmcv. Is there a reason that we remove it?

Comment (sxjscience) [reply .github/workflow_scripts/env_setup.sh:61]: It is causing CI to fail.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 9]
```diff
@@ -1771,6 +1875,8 @@ def predict(
         elif self._pipeline == OCR_TEXT_RECOGNITION:
             ret_type = [TEXT, SCORE]
 
+        pl.seed_everything(seed, workers=True)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1878]: Why do we need seeding in prediction?

File: .github/workflow_scripts/env_setup.sh (hunk 8 above)
Comment (sxjscience) [reply .github/workflow_scripts/env_setup.sh:61]: May be @Raldir we can try --timeout 1000.

Comment (sxjscience) [reply .github/workflow_scripts/env_setup.sh:61]: 120 can still be too small I think. May be we can also revise it in the next PR.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 9 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:1878]: I think it’s a good practice to seed before running models, even in the inference script. This avoids potential unreproducibility issues. An example is https://github.com/NVIDIA/FasterTransformer/blob/bc077a9078b99244c69cfe0dc44af86fd974bc71/examples/pytorch/bert/run_glue.py#L290

Comment (Raldir) [reply multimodal/src/autogluon/multimodal/predictor.py:1878]: Specifically T-Few selects a template randomly (both for training and inference as per default in the paper), so not seeding before running it results in reproducibility issues.

File: .github/workflow_scripts/env_setup.sh (hunk 8 above)
Comment (Raldir) [reply .github/workflow_scripts/env_setup.sh:61]: Ok, will revise to 1000

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 10]
```diff
@@ -1346,6 +1412,37 @@ def _default_predict(
...
+                lr=self._config.optimization.learning_rate,
+                lr_decay=self._config.optimization.lr_decay,
+                end_lr=self._config.optimization.end_lr,
+                lr_mult=self._config.optimization.lr_mult,
+                weight_decay=self._config.optimization.weight_decay,
+                warmup_steps=self._config.optimization.warmup_steps,
+            )
+        else:
+            optimization_kwargs = {}
+            trainable_param_names = []
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1444]: Do we still need this deepspeed strategy in inference? Isn't it only necessary for training?

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1444]: Consider adding a TODO to improve the design. Using `optimization_kargs` and `trainable_param_names` in inference is kind of confusing since no training and optimization are not expected in inference.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 9 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1878]: Here we only seed in `.predict()`. It can't be used by other APIs, e.g., `.predict_proba()`.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 11]
```diff
@@ -1560,6 +1662,7 @@ def _predict(
         data: Union[pd.DataFrame, dict, list],
         requires_label: bool,
         realtime: Optional[bool] = None,
+        seed=123,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1665]: The seed seems unused in `_predict()`?

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 12]
```diff
@@ -2034,9 +2147,11 @@ def save(self, path: str, standalone: Optional[bool] = True):
                 fp,
                 ensure_ascii=True,
             )
+
         # In case that users save to a path, which is not the original save_path.
         if os.path.abspath(path) != os.path.abspath(self._save_path):
             model_path = os.path.join(self._save_path, "model.ckpt")
+            print(model_path, path)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2154]: Remove `print()` or use `logger.debug()`.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 13]
```diff
@@ -1321,7 +1373,21 @@ def _top_k_average(
...
+                checkpoint = {
+                    "state_dict": {
+                        name.partition("module.")[2]: param
+                        for name, param in strategy.model._zero3_consolidated_16bit_state_dict().items()
+                    }
+                }
+            else:
+                checkpoint = {
+                    "state_dict": {"model." + name: param for name, param in self._model.state_dict().items()}
+                }
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1389]: I guess this is the same as `checkpoint = {"state_dict": avg_state_dict}` in `if not standalone`, right?

**Review by zhiqiangdon — APPROVED** [review]
Thanks for the awesome feature! Left several minor comments.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 12 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:2154]: @Raldir I think we can change it.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 13 above)
Comment (Raldir) [reply multimodal/src/autogluon/multimodal/predictor.py:1389]: In the situation that not all parameters are updated, `avg_state_dict` does not contain all model weights, only the updated ones. If `standalone==True`, we want to save entire the model's weights, hence going through the model's entire `state_dict` in that case.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 9 above)
Comment (Raldir) [reply multimodal/src/autogluon/multimodal/predictor.py:1878]: Ah right, good point!

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 10 above)
Comment (Raldir) [reply multimodal/src/autogluon/multimodal/predictor.py]: Good point, added a TODO now. Unfortunately, specifying `optmization_kwargs` is currently a requirement in pytorch-lightning integration of deepspeed but should be removed as soon as possible. Calling deepspeed still needed in inference to run large models to reduce computation to `largest_layer_memory + 2 * params`.
